src/contracts/orderbook/NiftyswapOrderbook.sol

```diff
@@ -0,0 +1,422 @@
+        }
+        address tokenContract = listing.tokenContract;
+
+        // Refund some gas
+        delete orders[listingId];
+
+        emit ListingCancelled(listingId, tokenContract);
+    }
+
+    /**
+     * Cancels an offer.
+     * @param offerId The ID of the offer.
+     */
+    function cancelOffer(bytes32 offerId) external {
+        Order storage offer = orders[offerId];
+        if (offer.creator != msg.sender) {
+            revert InvalidOfferId(offerId);
+        }
+        address tokenContract = offer.tokenContract;
+
+        // Refund some gas
+        delete orders[offerId];
+
+        emit OfferCancelled(offerId, tokenContract);
+    }
+
+    /**
+     * Deterministically create the orderId for the given order.
+     * @param order The order.
+     * @return orderId The ID of the order.
+     */
+    function hashOrder(Order memory order) public pure returns (bytes32 orderId) {
+        return keccak256(
+            abi.encodePacked(
+                order.creator,
+                order.tokenContract,
+                order.tokenId,
+                order.quantity,
+                order.currency,
+                order.pricePerToken,
+                order.expiry
+            )
+        );
+    }
+
+    /**
+     * Gets an order.
+     * @param orderId The ID of the listing or offer.
+     * @return order The order.
+     */
+    function getOrder(bytes32 orderId) external view returns (Order memory order) {
+        return orders[orderId];
+    }
+
+    /**
+     * Checks if orders are valid.
+     * @param orderIds The IDs of the orders.
+     * @return valid The validities of the orders.
+     * @notice An order is valid if it is active, has not expired and tokens (currency for offers, tokens for listings) are transferrable.
+     */
+    function isOrderValid(bytes32[] memory orderIds) external view returns (bool[] memory valid) {
+        valid = new bool[](orderIds.length);
+        for (uint256 i; i < orderIds.length; i++) {
+            Order memory order = orders[orderIds[i]];
+            valid[i] = order.creator != address(0) && !_isExpired(order)
+                && _hasApprovedTokens(order.tokenContract, order.tokenId, order.quantity, order.creator);
+        }
+    }
+
+    /**
+     * Checks if a order has expired.
+     * @param order The order to check.
+     * @return isExpired True if the order has expired.
+     */
+    function _isExpired(Order memory order) internal view returns (bool isExpired) {
+        // solhint-disable-next-line not-rely-on-time
+        return order.expiry <= block.timestamp;
+    }
+
+    /**
+     * Checks if a token contract is ERC1155 or ERC721.
+     * @param tokenContract The address of the token contract.
+     * @return isERC1155 True if the token contract is ERC1155, false if ERC721.
+     * @dev Throws if the token contract is not ERC1155 or ERC721.
+     */
+    function _tokenIsERC1155(address tokenContract) internal view returns (bool isERC1155) {
```

Comment:
I have the feeling that this is going to give us trouble with some tokens, imho it's a lot more sensible (in this context) to put a flag on the order struct, it could be packed alongside one of the addresses so it would have practically zero cost.

Reply:
Removed. It's now an input param